### ISSUE TITLE
actriverecord 5.2.3 breaks on numeric searches for empty strings

### ISSUE THREAD

**Issue by cork** [rails/rails#36413]
### Steps to reproduce
* create a mysql database units with a numeric column (int or bigint) tag
* run in `Unit.where(tag: "").to_sql` in a console
* run in `Unit.where.not(tag: "").to_sql` in a console

### Expected behavior
5.2.2.1 returns "SELECT `units`.* FROM `units` WHERE `units`.`tag` IS NULL"
5.2.2.1 returns "SELECT `units`.* FROM `units` WHERE `units`.`tag` IS NOT NULL"

### Actual behavior
5.2.3 returns "SELECT `units`.* FROM `units` WHERE `units`.`tag` = NULL"
5.2.3 returns "SELECT `units`.* FROM `units` WHERE `units`.`tag` != NULL"

### System configuration
**Rails version**:
5.2.3 and 5.2.2.1 or older
**Ruby version**:
2.6.x

## Comments
This makes form searches from web forms impossible for "empty values".

**Comment by NotGrm** [comment]
I also have this problem since I updated activerecord in version 5.2.3

## System configuration

Activerecord version:
5.2.3 
Ruby version:
2.5.x
MySQL version:
5.7.x

**Comment by michaelandhsm2** [comment]
Hi!

Looking at the tests written for the ActiveRecord library, I think that this behavior is intended.

https://github.com/rails/rails/blob/0aa86534d726090bc244b21367adc333ab961835/activerecord/test/cases/arel/attributes/attribute_test.rb#L409-L416

As you can see, the query method handles nil to produce your desired outcome and does not differentiate between numerical/non-numerical columns. You can replace the empty strings with nil by using Ruby's [presence](https://apidock.com/rails/Object/presence) method.

```
Unit.where(tag: "".presence).to_sql
=> SELECT "units".* FROM "units" WHERE "units"."tag" IS NULL

Unit.where(tag: "".presence).to_sql
=> SELECT "units".* FROM "units" WHERE "units"."tag" IS NOT NULL
```

**Comment by cork** [comment]
I agree with you that the test suite doesn't test it, but I don't agree that it is any proof that it is as intended.

**Comment by eileencodes** [comment]
@cork can you write a script [using these templates](https://github.com/rails/rails/tree/master/guides/bug_report_templates) that reproduce the error? That makes it easier for us to bisect and test the problem you're seeing.

**Comment by cork** [comment]
@eileencodes definitely, what do you mean by script, construct a rails install or what exactly?

**Comment by eileencodes** [comment]
Edit this template https://github.com/rails/rails/tree/master/guides/bug_report_templates to reproduce the error.

**Comment by eileencodes** [comment]
Oops https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb I mean this one.

**Comment by eileencodes** [comment]
Here's an example of one https://github.com/rails/rails/issues/36453 in another PR

**Comment by cork** [comment]
@eileencodes hope this helps.

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.integer :tag, null: true
  end
end

class Post < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_sql_queries
    assert_equal "SELECT \"posts\".* FROM \"posts\" WHERE \"posts\".\"tag\" IS NULL", Post.where(tag: nil).to_sql
    assert_equal "SELECT \"posts\".* FROM \"posts\" WHERE \"posts\".\"tag\" IS NULL", Post.where(tag: "").to_sql
  end
end
```

**Comment by kamipo** [comment]
hmm... actually the expected behavior in this issue (`"units"."tag" IS NULL`) is a regression for 5.2 #33624, that is different from 5.1.
#33624 is fixed by #35310, now 5.2.3 behaves the same with 5.1.

**Comment by cork** [comment]
Any movements on this?

To call it a fix to break if it gets form values without the controller filtering values out to not get broken sql queries is very backwards, else why isn't the same required to convert 0/1 to false/true for booleans too?

It is basically the same concept...

**Comment by cork** [comment]
If someone else ends up with the bug here is a workaround until someone wants to look at active record. Just drop in in config/initializers:
`class ActiveRecord::ConnectionAdapters::AbstractMysqlAdapter
  alias execute_without_null_fix execute
  def execute(sql, name = nil)
    execute_without_null_fix(sql.gsub(/` \!= NULL/, "` IS NOT NULL").gsub(/` = NULL/, "` IS NULL"), name)
  end
end
`
